feat(test runner): improve sharding algorithm to better spread similar tests among shards

muhqu opened the pull request:
Adds alternative algorithms to assign test groups to shards to better distribute tests.

## Problem

Currently the way sharding works is something like this…
```
         [  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12]
Shard 1:  ^---------^                                      : [  1, 2, 3 ]
Shard 2:              ^---------^                          : [  4, 5, 6 ]
Shard 3:                          ^---------^              : [  7, 8, 9 ]
Shard 4:                                      ^---------^  : [ 10,11,12 ]
```

Tests are ordered in the way they are discovered, which is mostly alphabetically. This has the effect that test cases are sorted nearby similar tests… for example your have first 6 tests which are testing logged-in state and then 6 tests which test logged-out state. The first 6 tests require more setup time as they are testing logged-in behaviour… With the current sharding algorithm shard 1 & 2 get those slow logged-in tests and shard 3 & 4 get the more quicker tests… 


## Solution

This PR adds a new `shardingMode` configuration which allows to specify the sharding algorithm to be used…

#### `shardingMode: 'partition'`

That's the current behaviour, which is the default. Let me know if you have a better name to describe the current algorithm...

#### `shardingMode: 'round-robin'`
Distribute the test groups more evenly. It…
1. sorts test groups by number of tests in descending order
2. then loops through the test groups and assigns them to the shard with the lowest number of tests.

Here is a simple example where every test group represents a single test (e.g. `--fully-parallel`) ...

```
         [  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12]
Shard 1:    ^               ^               ^              : [  1, 5, 9 ]
Shard 2:        ^               ^               ^          : [  2, 6,10 ]
Shard 3:            ^               ^               ^      : [  3, 7,11 ]
Shard 4:                ^               ^               ^  : [  4, 8,12 ]
```
…or a more complex scenario where test groups have different number of tests…
```
Original Order: [ [1], [2, 3], [4, 5, 6], [7], [8], [9, 10], [11], [12] ]
Sorted Order:   [ [4, 5, 6], [2, 3], [9, 10], [1], [7], [8], [11], [12] ]
Shard 1:           ^-----^                                                : [ [ 4,   5,   6] ]
Shard 2:                      ^--^                       ^                : [ [ 2,  3],  [8] ]
Shard 3:                              ^---^                    ^          : [ [ 9, 10], [11] ]
Shard 4:                                       ^    ^                ^    : [ [1], [7], [12] ]
```

#### `shardingMode: 'duration-round-robin'`

It's very similar to `round-robin`, but it uses the duration of a tests previous run as cost factor. The duration will be read from `.last-run.json` when available. When a test can not be found in `.last-run.json` it will use the average duration of available tests. When no last run info is available, the behaviour would be identical to `round-robin`.

## Other changes

* Add `testDurations?: { [testId: string]: number }` to `.last-run.json`
* Add builtin `lastrun` reporter, which allows `merge-reports` to generate a `.last-run.json` to be generated

## Appendix

Below are some runtime stats from a project I've been working on, which shows the potential benefit of this change.

The tests runs had to complete 161 tests. Single test duration ranges from a few seconds to over 2 minutes. 

<img width="361" align="left" alt="image" src="https://github.com/microsoft/playwright/assets/26716/621cd21e-c699-42d7-ba1f-1ffdf59569da">

The **partition** run gives the baseline performance and illustrates the problem quite good. We have a single shard that takes almost 16 min while another one completes in under 5 min. 

<br clear="both">

<img width="361" align="left" alt="image" src="https://github.com/microsoft/playwright/assets/26716/29dc5be2-6440-4e5f-8bb3-d8bd38e44600">

The **round-robin** algorithm gives a bit better performance, but it still has a shard that requires twice the time of another shard. 

<br clear="both">

<img width="362" align="left" alt="image" src="https://github.com/microsoft/playwright/assets/26716/779743d6-6290-4072-a8c3-ef50e70b1238">

The **duration-round-robin** run was using the duration info from a previous run and achieves the best result by far. All shards complete in 10-11 minutes. 🏆 🎉 

<br clear="both">


### Want to try this out, but don't know how? 

📦  [Here is a patch](https://gist.github.com/muhqu/04b81cf6b02850dad8124c82818ef154) for playwright 1.56.1 with the changes from this PR...
📦  [Here is a patch](https://gist.github.com/muhqu/f09987838ce8117d295e77dd3986cb98) for playwright 1.54.1 with the changes from this PR...
📦  [Here is a patch](https://gist.github.com/muhqu/037477132ce7cbd208d13bbb84edfc7e) for playwright 1.48.0 with the changes from this PR...
📦  [Here is a patch](https://gist.github.com/muhqu/4901f6830e3dc555aa6187013d89fe5e) for playwright 1.47.0 with the changes from this PR...

muhqu committed:
sharding algorithm to better spread similar tests among shards

muhqu commented:
Maybe it's better to make this an option to allow restoring the old behaviour. ¯\_(ツ)_/¯

And… there should be unit-tests, no?

muhqu committed:
improve shard algorithm by sorting test groups by number of tests

muhqu committed:
adjust shard.spec.ts to new algorithm

muhqu committed:
fix reporter-json.spec.ts due to sharding

muhqu committed:
fix reporter-junit.spec.ts due to sharding

muhqu committed:
empty commit to trigger ci

pavelfeldman commented:
Do you think you can achieve the same better behavior with your sharding seed? Or are you looking for additional bias against subsequent tests being put into the same group?

muhqu commented:
> Do you think you can achieve the same better behavior with your sharding seed?
Not sure yet. But I will test this new sharding logic in our test setup to gather some results.

> Or are you looking for additional bias against subsequent tests being put into the same group?
The seeded shuffle is basically just a quick and easy way to influence the test group to shard assignment… it's random and so it's results may vary.

However, this change is aimed to improve the sharding logic to generally yield better results, which yet needs to be proved. 😅

Currently this sharding algorithm uses the number of tests per test group as a cost metric. It would be great if we could use the test duration of a previous run (when available) to even better distribute the tests among the shards. But the algorithm would be quite similar.

pavelfeldman commented:
I think your seed change allows users to experiment with the seeds and arrive at some better state than they are today. Any other changes w/o the timing feedback are going to yield similar results, not need to experiment with biases.

> It would be great if we could use the test duration of a previous run (when available) to even better distribute the tests among the shards.

This requires a feedback loop with the test time stats which we don't have today. We recently started storing the last run stats in `.last-run.json`, I think it is Ok to store the test times there and to use it on the subsequent runs for better sharding, if it is available. Would you be interested in working on it?

muhqu commented:
Yes, I would like to work on that.

I was not yet aware of the .last-run.json. Is that something that is also written by the merge reports command? Because we need the stats combined from all shard runs.

I was thinkings about adding a separate reporter for that purpose, but if those last run stats are already there…, then there might not be the need to create a separate reporter.

muhqu committed:
add test durations to last run info

muhqu committed:
allow .last-run.json to be generated via merge-reports

muhqu commented:
I've added a `lastrun` reporter that can be used with merge-reports to generate `.last-run.json`.

Surprisingly when merging the reports the test ids just had a 1 character suffix that I was able to strip off… but it doesn't feel like the right way to do this.

What's the reason to modify test ids when merging blobs? Couldn't this be done in a way that only modifies a test id when there is a collision?

muhqu committed:
Add shardingMode configuration

muhqu committed:
revert json/junit test changes

muhqu committed:
empty commit to trigger ci

muhqu commented on packages/playwright/src/runner/runner.ts:
```diff
@@ -161,7 +162,12 @@ async function writeLastRunInfo(testRun: TestRun, status: FullResult['status'])
   const lastRunReportFile = path.join(outputDir, '.last-run.json');
   const failedTests = testRun.rootSuite?.allTests().filter(t => !t.ok()).map(t => t.id);
-  const lastRunReport = JSON.stringify({ status, failedTests }, undefined, 2);
+  const testDurations = testRun.rootSuite?.allTests().reduce((map, t) => {
+    if (t.results.length)
+      map[t.id] = t.results.reduce((a, b) => a + b.duration, 0);
+    return map;
+  }, {} as { [testId: string]: number });
```
I'm actually not sure it is the right way to sum all the durations… maybe it makes more sense to calc the average? Or only include durations from successful test runs… 🤔

dgozman replied:
I think this is ok for now, we'll iterate based on feedback.

muhqu commented:
@pavelfeldman `last-run.json` is a little bit tricky to work with atm… it constantly gets overwritten even if you just list tests and VS Code seems to do that from time to time to refresh the tests panel... I think it should only get written when it's actually running tests.

muhqu committed:
fix(merge-reports) only change test ids when needed

muhqu committed:
fix(merge-reports) only change test ids when needed

muhqu committed:
remove special handling of merged test ids

muhqu committed:
Merge branch 'main' into sharding-algorithm

muhqu committed:
fix(runner) don't write last run info when listing tests

muhqu committed:
optimize: use single global set

muhqu committed:
Feedback use outputLines

muhqu committed:
Revert "fix(merge-reports) only change test ids when needed"

This reverts commit 389e5711f46f7bf6b0be98ca15795e17e4d3958b.

muhqu committed:
Merge branch 'fix-merge-reports-test-ids' into sharding-algorithm

muhqu committed:
empty commit to trigger ci

muhqu committed:
lint fix

muhqu committed:
Merge branch 'fix-merge-reports-test-ids' into sharding-algorithm

muhqu commented:
> @pavelfeldman […] we would probably explicitly point to the file […]

Do you have a recommendation how you would name cli parameter / configuration option? 

Something like `--sharding-read-last-run-info path/to/merged-last-run.json` is probably too much of a mouth full? However, I think we should somehow make it obvious when data is read or written to a file…

muhqu committed:
Merge branch 'main' into sharding-algorithm

lukeapage commented:
This is awesome - we were planning on implementing something similar outside playwright before we migrate cypress tests that currently shard across 50 build servers.

Is there anything todo here I could help with to get this merged quicker ?

muhqu commented:
Hey, I did not noticed this PR got conflicts due to the revert of my other change… I’ll resolve those conflicts in the next days. There was still on open question on whether we should introduce a command line flag to specify a last run state file or not… but I think that are improvements that can also happen after this has been merged

Though, this PR could have some more documentation changes… anyone want to help with that?

muhqu committed:
Merge branch 'main' into sharding-algorithm

muhqu committed:
fix test.d.ts

dgozman commented on packages/playwright/src/runner/runner.ts:
```diff
@@ -152,6 +152,7 @@ export class Runner {
 export type LastRunInfo = {
   status: FullResult['status'];
   failedTests: string[];
+  testDurations?: { [testId: string]: number };
```
We are a bit worried that this will force `testId` to never change to preserve backwards compatibility. Therefore, we would like to use `testCase.titlePath()` instead of id for now. Something like this:

```ts
export type LastRunInfo = {
  status: FullResult['status'],
  projects: {
    name: string,
    files: {
      path: string,
      tests: {
        title: string[],
        duration: number,
        status: TestStatus,
      }[],
    }[],
  }[],
};
```
 
Here `file.path` is `toPosixPath(path.relative(config.rootDir, filePath))`, similar to [this line](https://github.com/microsoft/playwright/blob/main/packages/playwright/src/common/suiteUtils.ts#L56).

muhqu replied:
I just tried to write the LastRunInfo in the proposed format, but its not as straight forward as I thought it would be.

> That said, you can leave it as is, and we'll iterate on it.

You mean, I don't need to proceed here to get this merged? 🙏

muhqu replied:
> […] we do not expect last-run.json to be used between different versions

In our CI setup, I'm using pipeline steps to restore the last-run.json from the last successful main build. So, in our case last-run info from previous playwright versions will be used, and it would be great if there is some level of backward compatibility... at least it should not fail when the format is incorrect.

dgozman reviewed (commented):
> There was still on open question on whether we should introduce a command line flag to specify a last run state file or not

We discussed this and decided that it should be fine to add `--last-run-file=` command line flag that will affect both the reading and writing location of the `last-run.json`.

> Though, this PR could have some more documentation changes… anyone want to help with that?

Of course! We can help with adding more documentation after landing this PR.

muhqu committed:
add --last-run-file CLI parameter

muhqu commented:
@dgozman I've just added `--last-run-file=` cli flag and CI seems to be happy… any further work needed on this PR to get merged?

muhqu committed:
Adjust since to v1.46

muhqu committed:
Merge branch 'main' into sharding-algorithm

muhqu committed:
Add documentation on sharding modes

muhqu committed:
linting

muhqu committed:
Add into text to more complex round-robin scenario

muhqu committed:
no ts highlight for sharding illustration code blocks

muhqu committed:
use yaml highlight for sharding illustration code blocks as NO highlighting is not an option

muhqu committed:
revert / reorder import changes

muhqu commented:
Just added some more documentation, mostly copying what’s in this PR description to the test-sharding page. And cleaned up imports which were previously automatically “organized” due to my vs code settings, which resulted in unnecessary changes with high conflict potential…

It would be great if we could get this over the finish line for the next playwright release. People already asked me how they could use playwright of my PR in their CI… 😅

@dgozman @pavelfeldman can I get a review?

dgozman reviewed (commented):
@muhqu Sorry for the slow review, I was unavailable for a week.

I think we can land this PR if you remove the reporter for now, and remove the `round-robin` strategy due to stability concerns. This will give us the necessary functionality, and we can iterate further.

I can follow up with updating the `duration-round-robin` to fallback to `partition`.

dgozman commented on docs/src/test-api/class-testconfig.md:
```diff
@@ -482,6 +482,27 @@ export default defineConfig({
+Defines the algorithm to be used for sharding. Defaults to `'partition'`.
+* `'partition'` - divide the set of test groups by number of shards. e.g. first
+  half goes to shard 1/2 and seconds half to shard 2/2.
+* `'round-robin'` - spread test groups to shards in a round-robin way. e.g. loop
```
I guess that's fair. If you would like to opt-in into `round-robin` and it works for you, that's great. I still think it is less stable than I would prefer, but it doesn't hurt to have it as an option anyway.

muhqu commented on docs/src/test-api/class-testconfig.md:
```diff
@@ -482,6 +482,27 @@ export default defineConfig({
+  test groups to shards in a round-robin way. e.g. loop over test groups and
+  always assign to the shard that has the lowest duration of tests. new tests
+  which were not present in the last run will use an average duration time. When
+  no `.last-run.json` could be found the behavior is identical to
```
I disagree.

Also please note that `duration-round-robin` and 'round-robin' are using the same implementation. The only difference is that in case of 'duration-round-robin' the last run info is passed to the implementation. However, in case the last run info is passed but it is empty, the behaviour is identical to 'round-robin' where no last run info is passed.

I don't think it would be a good to choose a different algorithm based on whether last run info is available or not.

muhqu commented on packages/playwright/src/runner/runner.ts:
```diff
@@ -152,17 +152,24 @@ export class Runner {
+      map[t.id] = t.results.reduce((a, b) => a + b.duration, 0);
+    return map;
+  }, {} as { [testId: string]: number });
+  const lastRunReport = JSON.stringify({ status, failedTests, testDurations }, undefined, 2);
```
The `lastrun` reporter is required to generate last run info when merging blob reports from shards.

Maybe it would be better to implicitly write the lastrun info when merging blob reports?

dgozman replied:
I think so. From the user's perspective, lastrun just working out of the box would be easier. Whether it is a reporter internally or not is an implementation detail.

muhqu replied:
How about always using the last run reporter? Then we would only need to have a single implementation….

dgozman replied:
That definitely works. We'll have to add it [here](https://github.com/microsoft/playwright/blob/main/packages/playwright/src/runner/reporters.ts#L37) when `isTestServer === false`.

alexmalev commented:
really looking forward to this one! thanks for the good work

muhqu commented:
Sorry, for the delay… I'll see if I can find some more time to work on this to get it over the finish line…

muhqu committed:
Merge branch 'main' into sharding-algorithm

muhqu commented on tests/playwright-test/shard-roundrobin.spec.ts:
```diff
@@ -0,0 +1,457 @@
+  ]);
+});
+
+test('should not produce skipped tests for zero-sized shards', async ({ runInlineTest }) => {
```
Okay, I'll remove the test which are not affected by sharding.

muhqu committed:
feedback

muhqu committed:
feedback - remove redundant tests

muhqu committed:
always use lastrun reporter and keep em internal

muhqu committed:
validate --sharding-mode CLI parameter

muhqu committed:
export type ShardingMode

muhqu committed:
fix .last-run.json location

muhqu commented on packages/playwright/src/runner/reporters.ts:
```diff
@@ -78,6 +77,13 @@ export async function createReporters(config: FullConfigInternal, mode: 'list' |
+
+  if (!isTestServer && (mode === 'test' || mode === 'merge')) {
+    // If we are not in the test server, always add a last-run reporter.
+    const lastRunOutputFile = config.projects.length >= 0 ? path.join(config.projects[0].project.outputDir, '.last-run.json') : undefined;
+    reporters.push(new LastRunReporter({ ...runOptions, outputFile: lastRunOutputFile }));
```
It kind of feels wrong that the `.last-run.json` is only written to the first project `outputDir`, but that's how it is at the moment.

muhqu committed:
fix .last-run.json location in reporter-lastrun spec

muhqu committed:
use lastRunFile from config when defined

muhqu committed:
Merge branch 'main' into sharding-algorithm

muhqu commented on packages/playwright/src/runner/reporters.ts:
```diff
@@ -76,6 +77,13 @@ export async function createReporters(config: FullConfigInternal, mode: 'list' |
+
+  if (!isTestServer && (mode === 'test' || mode === 'merge')) {
+    // If we are not in the test server, always add a last-run reporter.
+    const lastRunOutputFile = config.lastRunFile ?? config.projects.length >= 0 ? path.join(config.projects[0].project.outputDir, '.last-run.json') : undefined;
+    reporters.push(new LastRunReporter({ ...runOptions, outputFile: lastRunOutputFile }));
```
@dgozman It kind of feels wrong that the default location for `.last-run.json` is derived from the first project's `outputDir`, but that's how it works in `main` at the moment, so I choose to do it the same here...

muhqu committed:
Merge branch 'main' into sharding-algorithm

muhqu commented:
@alexmalev if you want to test the new sharding algorithm, [here is a patch](https://gist.github.com/muhqu/4901f6830e3dc555aa6187013d89fe5e) for playwright 1.47.0 with the changes from this PR. Let us know if it works for you and whether it improves overall runtime performance. 🙏

alexmalev commented:
> @alexmalev if you want to test the new sharding algorithm, [here is a patch](https://gist.github.com/muhqu/4901f6830e3dc555aa6187013d89fe5e) for playwright 1.47.0 with the changes from this PR. Let us know if it works for you and whether it improves overall runtime performance. 🙏  

Will do. Thanks!

muhqu commented:
@pavelfeldman @dgozman I've updated this sharding improvements PR. The failing test (_babel.spec.ts:135:5 › should not transform external_) doesn't seem to have something to do with the changes in this PR…

Open question: The data structure of `LastRunInfo`… I'm currently simply adding `testDurations?: { [testId: string]: number }` and @pavelfeldman mentioned it would be better to not rely on Test IDs to be consistent between runs and possibly playwright version updates… What alternatives do we have? Test title path does not need to be unique, right? …what about profile configs (e.g. test that run fast on chrome but slow on firefox)? 

If you ask me, I think we should be fine with using TestIds and kind of make it a goal to have them consistent between test runs (not necessarily between playwright upgrades, but at least between runs)… merging of reports with colliding TestIds is already taken care of since https://github.com/microsoft/playwright/pull/31061.

alexmalev commented:
> @alexmalev if you want to test the new sharding algorithm, [here is a patch](https://gist.github.com/muhqu/4901f6830e3dc555aa6187013d89fe5e) for playwright 1.47.0 with the changes from this PR. Let us know if it works for you and whether it improves overall runtime performance. 🙏

This is brilliant, we run 130 tests in full parallelism, using 4 shards. 
this is a typical run before the patch (partition sharding mode):
![image](https://github.com/user-attachments/assets/f667136a-5275-4ce9-b895-e6c2d5a12a0c)
and here is using your patch (in the naive round-robin sharding mode):
![image](https://github.com/user-attachments/assets/c2c47125-4f25-4b9c-b3ed-e13d69e7c4b3)

its so good that i wasnt even bothered with saving the last-run duration in the ci. 

good job!

muhqu commented:
There is still an issue with the `--last-run-file` command line flag I need to fix...

muhqu committed:
validate config.shardingMode

muhqu committed:
fix --last-run-file parameter

muhqu commented:
hm… okay, now there are a bunch of conflicts due to @dgozman 's  #32540 (_chore(test runner): extract LastRunReporter_)

I'll see if I can adapt to that…

muhqu committed:
Merge branch 'main' into sharding-algorithm

muhqu committed:
adapt LastRunReporter

muhqu committed:
use separate LastRunReporter instances

muhqu committed:
allow lastRun.ts to be imported by merge.ts

muhqu committed:
Merge branch 'main' into sharding-algorithm

muhqu commented:
Btw, what's the reason for `playwright merge-reports` command to be [hidden](https://github.com/microsoft/playwright/pull/22509/files)? I mean why is it not listed when running `playwright --help`?
\cc @yury-s @aslushnikov

muhqu committed:
Add test.skip for Node.js bug in 'should not transform external' babel.spec.ts

muhqu commented:
Here is a separate PR to silence the `babel.spec.ts` node.js issue : https://github.com/microsoft/playwright/pull/32604

muhqu committed:
Merge branch 'main' into sharding-algorithm

muhqu committed:
Revert "Add test.skip for Node.js bug in 'should not transform external' babel.spec.ts"

This reverts commit 412e05bc8d139ece8f23c2b9fd8d3dc2e51cba15.

liviucmg commented:
@muhqu We have about 400 tests split into 32 shards. Here's a sample before-and-after using `duration-round-robin`, in which the longest shard is reduced from ~23 minutes to ~16 minutes. It still fluctuates due to other factors, of course, like flaky tests, but overall amazing job! :clap: 

![image](https://github.com/user-attachments/assets/fb167f4a-78e3-451a-8036-406df92d2726)

muhqu committed:
Merge branch 'main' into sharding-algorithm

muhqu committed:
Merge branch 'main' into sharding-algorithm

muhqu committed:
Merge branch 'main' into sharding-algorithm

muhqu committed:
empty commit to trigger CI

muhqu commented:
@dgozman @pavelfeldman how is it? …do you have an idea when you get time to review this PR again? Would love to get this finished…

muhqu committed:
Merge branch 'main' into sharding-algorithm

ValorHeart commented:
This feature looks like a powerful capability, which will also distinguish Playwright from competitors who don't offer a solution to such a case.
I hope it'll get added soon.

muhqu commented:
@ValorHeart If you want to try this out, 📦 [Here is a patch](https://gist.github.com/muhqu/4901f6830e3dc555aa6187013d89fe5e) for playwright 1.47.0 with the changes from this PR...

pavelfeldman commented:
We've discussed it at length during the team meeting. The consensus was that we don't see a low-maintenance solution to the problem that would cover a meaningful number of use cases. Even the mature solutions that track execution timing and reuse this information in subsequent runs are brittle and have tendency to rot. We would go for a lower-level solution that would allow user to take over the scheduling (test list files or api calls), but we don't want to see a new extensive api surface or file formats there. I'm following up for transparency - we are thinking about the problem, but we don't see a good maintainable solution atm.

pavelfeldman commented:
Let me close this PR as the problem is not in the realm of the code, we should open an issue and have this discussion there first to commit to a solution.

muhqu commented:
@pavelfeldman that’s unfortunate and I don’t quite understand what you mean with “maintenance” in this context? How is the solution in this PR increasing maintenance cost? 

Will you create an issue or are you expecting me to create one?

GuskiS commented:
Was looking forward to this PR 😞 
![image](https://github.com/user-attachments/assets/3e58e938-6207-418c-9100-f2e31d7b1093)

liviucmg commented:
The MR is quite minimal in my opinion and does its job well enough. Also, it was easy for me to just save the `.last-run.json` file in an S3 bucket and retrieve it on subsequent runs.

We've been struggling with this problem for a long time and there are no other solutions that I'm aware of, other than one or two very expensive 3rd party services.

lukeapage commented:
This is a shame. We have a setup ready for this and was just awaiting a merge.

now we need to figure it out ourselves and then manually overwrite the config file with a list of tests to run on each shard.

lukeapage commented:
Btw we are coming from cypress where we wrote our own scheduler - we have 25 hours of tests and run them all hundreds of times a day - the last run result from master is very sufficient and does not “rot”

muhqu commented:
📦 [Here is a patch](https://gist.github.com/muhqu/037477132ce7cbd208d13bbb84edfc7e) for playwright 1.48.0 with the changes from this PR...

muhqu commented:
FYI @pavelfeldman @dgozman I've created a test PR (https://github.com/microsoft/playwright/pull/33032) to show how the round-robin sharding can improve playwrights own CI time… 
![image](https://github.com/user-attachments/assets/eb0e8ac8-d2a9-4c9b-bdee-6c406fd70cfd)

muhqu commented:
FYI, after a conversation with @pavelfeldman on #33032, I'm exploreing to add a test filter which should allow playwright users to create their own sharding logic #33049 … let's see how that goes…

muhqu commented:
I like to revive discussion on how to improve playwright sharding using timing data: https://github.com/microsoft/playwright/issues/17969#issuecomment-2421653419

arishoham commented:
Would anyone have a new patch for the latest version of playwright?

muhqu commented:
I will create a new patch in a few weeks... it's on my todo list.

muhqu commented:
@arishoham thanks for raising this… actually it had been on my todo list for too long, so I sat down and ported the changes to latest main.

My [`sharding-algorithm`](https://github.com/muhqu/playwright/tree/sharding-algorithm) branch is now back up to date with main and I've [created a patch against playwright 1.54.1](https://gist.github.com/muhqu/f09987838ce8117d295e77dd3986cb98). If you're trying it out, let me know how it goes…
